src/estimagic/optimization/tranquilo/sample_points.py

```diff
@@ -110,47 +117,207 @@ def _naive_sampler(
     rng,
+    ord,  # noqa: A002
     existing_xs=None,
     bounds=None,
+    algorithm="scipy_lbfgsb",
+    algo_options=None,
 ):
-    n_points = _get_effective_n_points(target_size, existing_xs)
+    """Optimal generation of trustregion points on the hull of general sphere / cube.
+
+    Points are sampled optimally on a hull (of a sphere for ord=2 and of a cube for
+    ord=np.inf), where the criterion that is maximized is the (smooth) minimum distance
+    of all pairs of points, except for pairs of existing points. These points are then
+    mapped into the feasible region, which is defined by the intersection of the
+    trustregion and the bounds.
+
+    Args:
+        trustregion (TrustRegion): NamedTuple with attributes center and radius.
+        target_size (int): Target number of points in the combined sample of existing_xs
+            and newly sampled points. The sampler does not have to guarantee that this
+            number will actually be reached.
+        rng (numpy.random.Generator): Random number generator.
+        ord (int): Type of norm to use when scaling the sampled points. For 2 it will
+            result in sphere sampling, for np.inf in cube sampling.
+        existing_xs (np.ndarray or None): 2d numpy array in which each row is an
+            x vector at which the criterion function has already been evaluated, that
+            satisfies lower_bounds <= existing_xs <= upper_bounds.
+        bounds (Bounds or None): NamedTuple.
+        algorithm (str): Optimization algorithm.
+        algo_options (dict): Algorithm specific configuration of the optimization. See
+            :ref:`list_of_algorithms` for supported options of each algorithm. Default
+            sets ``stopping_max_iterations=3``.
+
+    Returns:
+        np.ndarray: Generated points. Has shape (target_size, len(trustregion.center)).
+
+    """
+    n_points = _get_effective_n_points(target_size, existing_xs=existing_xs)
     n_params = len(trustregion.center)
-    raw = rng.normal(size=(n_points, n_params))
-    denom = np.linalg.norm(raw, axis=1).reshape(-1, 1)
 
-    points = trustregion.radius * raw / denom + trustregion.center
+    if n_points == 0:
+        return np.array([])
 
-    if bounds is not None and (bounds.lower is not None or bounds.upper is not None):
-        bounds = _get_effective_bounds(trustregion, bounds)
-        points = np.clip(
-            points,
-            a_min=bounds.lower,
-            a_max=bounds.upper,
-        )
+    if algo_options is None:
+        algo_options = {"stopping_max_iterations": 3}
 
+    effective_bounds = _get_effective_bounds(trustregion, bounds=bounds)
+
+    if existing_xs is not None:
+        # map existing points into unit space for easier optimization
+        existing_xs = (existing_xs - trustregion.center) / trustregion.radius
+
+    # start params
+    x0 = rng.normal(size=(n_points, n_params))
+    x0 = _project_onto_unit_hull(x0, ord=ord)
+
+    res = em.maximize(
```

Comment:
I would use scipy minimize to save the overhead. Since you use a 2d numpy array as params we don't even take the fast path and treat params as pytree.

Reply:
Very good point. For now, I'm sticking with estimagic but will use 1d-arrays in the optimization to use the fast path. From here on it will be easy to switch to `scipy.minimize` if we decide to do that at some point.